### ISSUE TITLE
Batchnorm

### ISSUE THREAD

**Issue by purzelrakete** [feldberlin/wavenet#18]
## What

Test the effect of Batch Normalisation on test / train likelihood and training times. 

## Hypothesis

The current network is having some stability issues. This is likely due to a poor initialisation strategy, which is investigated in #11. 

A different strategy is to batch normalise the network. Batch normed networks are able to overcome poor initialisation, and often reach better accuracy and training times than careful initialisation alone. 

Consequently Batch norm

- is a good baseline for what might be possible in batch norm free methods such as weight norm. 
- might be a viable alternative to those methods. 

Note that no other public Wavenet implementations uses Batch norm. I don't immediately see a reason why Batch norm should be inappropriate for a Wavenet, but need to think this through properly. 

## Results

Write up the results of your experiment once it has completed and has been analysed. Include links to the treatment run, and also to the baseline if appropriate.

## Acceptance Criteria

- [x] BN is applied before the nonlinearities
- [x] BN can be toggled on and off
- [x] BN implemented inside custom modules. Introduce one for the 1x1s
- [x] Model size adjusted to run on 2x 2080ti
- [x] Run a baseline with BN off, same model size. Tune to an optimal non crashing learning rate and then run the basline experiment to completion (2 epochs)
- [x] Run treatment with tuned learning rate. Should be much higher. Run treatment to completion.
- [x] Normalise the embedding layer too
- [x] Check that gammas and betas are applied per feature map, not per activation
- [x] Add BN to final b1x1
- [x] Remove bias terms in the convolutions when BN is active. Try to increase the batch size since we saved some memory
- [x] Unit test that Generator and Wavenet are producing the same results
- [x] Make sure that Tiny runs through and looks sane, with and without BN
- [x] Make sure that Sines runs through and looks same, with and without BN
- [x] Experiment is written up and linked from somewhere

**Comment by purzelrakete** [comment]
## Analysis

I added BatchNorm as a feature flag to the config. Running BatchNorm on Tiny and Sines shows a somewhat worse outcome when using a random decoder. I don't know what this should be the case. 

There's a possibility that I introduced an error into the version which runs when batch norm is  turned off. It's hard to say since I'm just eyeballing results in Tiny and Sines. It would be helpful to have some metrics for those datasets beyond just likelihood scores.
